compiler/rustc_feature/src/active.rs

```diff
@@ -547,6 +549,13 @@ declare_features! (
 /// Some features are not allowed to be used together at the same time, if
 /// the two are present, produce an error.
 ///
-/// Currently empty, but we will probably need this again in the future,
-/// so let's keep it in for now.
-pub const INCOMPATIBLE_FEATURES: &[(Symbol, Symbol)] = &[];
+/// This is also used for features which, whilst supposed to be *used* (_e.g._, by
+/// macros using `#[rustc_allow_internal_unstable(…)]`), are not supposed to be
+/// *globally enabled* explicitly. Indeed, allowing such a thing would not only
+/// be unnecessary (internal `unsafe` feature), it would also be unsound.
+/// The error message on misusage that this hack produces may not be pretty,
+/// but at least it is sound.
+pub const INCOMPATIBLE_FEATURES: &[(Symbol, Symbol)] = &[
+    // Would otherwise allow non-`unsafe` —and thus, unsound— access to `Pin` constructions.
+    (sym::unsafe_pin_internals, sym::unsafe_pin_internals),
+];
```

Comment:
I don't think we should do this. We usually don't touch `rustc_feature` at all for unstable library features, and there are already plenty of unstable features that could be used for unsoundness when enabled. We shouldn't try to attach any soundness/safety guarantee to unstable features, as that just raises the expectations of unstable features. The feature being permanently unstable, having `unsafe` in the name, and having `internals` in the name, is more than enough.

(Also, this produces a weird error about a feature being incompatible with itself.)

Reply:
> there are already plenty of unstable features that could be used for unsoundness when enabled

There are? Interesting. FWIW, as far as I know, for *language* features soundness bugs even in unstable features are considered a problem, and such features are typically listed as "incomplete feature". On the language level, it is *not* considered okay to have unsoundness when enabling a feature gate.

Looks like T-libs-api and T-lang have different policies here?

Reply:
I remember that a few things around specialization resulted in unsoundness in the standard library. While it did involve an attribute with `unsafe_` in the name and some code making assumptions about it, these do seem to be features where the soundness story isn't fully understood yet, without these features being marked as 'incomplete'.

I'm also reminded of the (recently removed) unstable `std::io::Initializer`, which entire purpose was to simplify working with uninitialized `&mut [u8]` buffers, resulting in unsoundness. Cases like these are a lot more subtle than the field in `Pin` of course, and it still involves `unsafe` functions.

Recently we also had `Pin::static_ref` as `unstable`. Before our discussion and the FCP blessing that, that was also technically unsound as you argued on that tracking issue.

(It'd be nice if we could mark the *field* of `Pin` as unsafe, but totally not worth it for just a single use case. In almost all use cases, an `unsafe` constructor function is fine.)

All these cases are much more subtle than the one in this PR though. Maybe the unsound unstable features situation isn't as bad as I thought. Still, I would still want to be able to experiment with unstable features in the library for which we're not entirely sure yet what the soundness story is (as long as we properly document that, of course).

Reply:
Would marking it as incomplete rather than this "incompatible with itself" be a decent middle-ground, then?

### The soundness contract

For me two things have to be distinguished <sub><sup>(note: I'll be talking of `incomplete_feature`s as the ones representing `potentially_unsound_feature`s (which would otherwise be a superset of it), since the latter do not exist yet. But feel free to imagine there being such `potentially_unsound_feature`s, in which case just replace my mentions to the former with mentions to the latter)</sup></sub>:

  - potentially unsound features, such as specialization, `std::io::Initializer`, _etc._,

    which I'd _personally_ like to be, at least, marked as `incomplete` (since the warning there does mention potential lack of soundness);

  - and the believed-to-be-sound features, even if they later appear to be unsound (at which point they'd either be removed, fixed, or would fall into the previous category). I'd put pre-FCPed-`Pin::static_ref`[ˆ1] in this category, for instance.

[^1]: Very interesting read, that was, btw!

Pros/cons of this approach:

  - :x: More maintenance burden for compiler / libs authors, which may look surprising when we are talking of `unstable`-gated stuff, whose purpose is quite tied to "avoiding" / minimizing that.

    I'd personally hypothesize this may not be that bad / be manageable, since we could be very lenient _w.r.t._ tagging features as `potentially_unsound`.

  - :white_check_mark: The benefit is more trust _w.r.t._ `nightly` APIs / less incentive to mistrust `nightly` altogether in highly-safety-concerned scenarios:

    assume a `#![forbid(unsafe_code)]` codebase (in a transitive fashion, with carefully reviewed allow-listed exceptions, such as the stdlib itself). With this approach, by adding a (similarly transitive) `#![forbid(incomplete_features)]`, they'd be able to "trust" the lack of UB in their codebase, _modulo bugs in the trusted dependencies_, such as the language itself, or the stdlib.

    By not banning `nightly` `#![feature]`s altogether, this codebase would still be able to use things such as [`Result::copied()`](https://doc.rust-lang.org/1.58.1/std/result/enum.Result.html#method.copied), **which is currently `unstable`, but harmless**: without a (lint-enabled) way to distinguish between `Result::copied` and access to the `.pointer` field of `Pin`, such codebases might have to maintain their own allow-list of the "harmless" `nightly` features, or _ban all of nightly altogether_.

Now, I really don't care what the distinguisher is, provided it can be linted against: `incomplete_features`, `potentially_unsound_features`, `internal_features` (maybe this last one expresses the right intent _w.r.t._ maintainability burden?).

In that regard, _naming_, in and of itself, is already a (slightly brittle) form of distinction: with `unsafe` and `internal` in the name, I agree that qualifying it as unsound sounds a bit silly / hypocrite (even if technically correct!).

### Back to this PR

That's why I'll settle either way, although I'd _personally_ lean towards, _at least_, keeping the feature tagged as `incomplete` (even if this technically makes it become a `lang` feature rather than a lib one).

Reply:
> I remember that a few things around specialization resulted in unsoundness in the standard library. While it did involve an attribute with unsafe_ in the name and some code making assumptions about it, these do seem to be features where the soundness story isn't fully understood yet, without these features being marked as 'incomplete'.

Specialization is marked incomplete for this exact reason, and min_specialization is sound to my knowledge (in particular since that attribute you mention is `unsafe`).

> Recently we also had Pin::static_ref as unstable. Before our discussion and the FCP blessing that, that was also technically unsound as you argued on that tracking issue.

I argued that we should not consider it sound, but it was only unsound if we had accepted that other example I gave there as sound. So that's a different case -- it is one of those cases in the gray area between sound and unsound (similar to e.g. how drop-based scoped threads are sound as long as Rc/Arc/mem::forget are considered unsound).

>  Still, I would still want to be able to experiment with unstable features in the library for which we're not entirely sure yet what the soundness story is (as long as we properly document that, of course).

Sure, no disagreement here. But for this feature here we *are* sure that it is unsound.

> Would marking it as incomplete rather than this "incompatible with itself" be a decent middle-ground, then?

Seems reasonable, for now (though the name "incomplete" seems to make less sense here).

Reply:
So, regarding https://github.com/rust-lang/rust/pull/93176/commits/8aa3a420ec2043f168f06146c0e0c26abc2bb1ef, for the moment I've settled for https://github.com/rust-lang/rust/pull/93176/commits/08ca99948837a9117bb85123f230cdfaadb83e87;

  - if the resulting situation is still deemed excessive, I can remove the latter from the history, and `git revert` the former. Note the difference: if we go with neither incompatible-with-itself nor `incomplete`, then I at least ask that an explicit removal of these things be present in the commit history.

  - otherwise, I can just squash those two commits down into one.

Reply:
Ah, looks like there are still bugs. :/
"It was thought to be sound", then. ;)